cpp/src/arrow/compute/exec/union_node.cc

```diff
@@ -0,0 +1,150 @@
+#include "arrow/compute/exec/options.h"
+#include "arrow/compute/exec/util.h"
+#include "arrow/util/bitmap_ops.h"
+#include "arrow/util/checked_cast.h"
+#include "arrow/util/future.h"
+#include "arrow/util/logging.h"
+#include "arrow/util/thread_pool.h"
+
+namespace arrow {
+
+using internal::checked_cast;
+
+namespace compute {
+
+namespace {
+std::vector<std::string> GetInputLabels(const ExecNode::NodeVector& inputs) {
+  std::vector<std::string> labels(inputs.size());
+  for (size_t i = 0; i < inputs.size(); i++) {
+    labels[i] = "input_" + std::to_string(i) + "_label";
+  }
+  return labels;
+}
+}  // namespace
+struct UnionNode : ExecNode {
+  UnionNode(ExecPlan* plan, std::vector<ExecNode*> inputs)
+      : ExecNode(plan, inputs, GetInputLabels(inputs),
+                 /*output_schema=*/inputs[0]->output_schema(),
+                 /*num_outputs=*/1) {
+    ARROW_DCHECK(this->input_count_.SetTotal(static_cast<int>(inputs.size())) == false);
+  }
+
+  const char* kind_name() override { return "UnionNode"; }
+
+  static Result<ExecNode*> Make(ExecPlan* plan, std::vector<ExecNode*> inputs,
+                                const ExecNodeOptions& options) {
+    RETURN_NOT_OK(ValidateExecNodeInputs(plan, inputs, static_cast<int>(inputs.size()),
+                                         "UnionNode"));
+    if (inputs.size() < 1) {
+      return Status::Invalid("Constructing a `UnionNode` with inputs size less than 1");
+    }
+    auto schema = inputs.at(0)->output_schema();
+    for (auto input : inputs) {
+      if (!input->output_schema()->Equals(schema)) {
+        return Status::Invalid(
+            "UnionNode input schemas must all match, first schema was: ", *schema,
+            " got schema: ", *input->output_schema());
+      }
+    }
+    return plan->EmplaceNode<UnionNode>(plan, std::move(inputs));
+  }
+
+  void InputReceived(ExecNode* input, int seq, ExecBatch batch) override {
+    ARROW_DCHECK(std::find_if(inputs_.begin(), inputs_.end(), [input](ExecNode* n) {
+                   return n == input;
+                 }) != inputs_.end());
+
+    if (finished_.is_finished()) {
+      return;
+    }
+    outputs_[0]->InputReceived(this, seq, std::move(batch));
+    if (batch_count_.Increment()) {
+      finished_.MarkFinished();
+    }
+  }
+
+  void ErrorReceived(ExecNode* input, Status error) override {
+    DCHECK_EQ(input, inputs_[0]);
+    outputs_[0]->ErrorReceived(this, std::move(error));
+
+    StopProducing();
+  }
+
+  void InputFinished(ExecNode* input, int num_total) override {
+    ARROW_DCHECK(std::find_if(inputs_.begin(), inputs_.end(), [input](ExecNode* n) {
+                   return n == input;
+                 }) != inputs_.end());
+
+    total_batches_.fetch_add(num_total);
+
+    if (input_count_.Increment()) {
+      outputs_[0]->InputFinished(this, total_batches_.load());
+      if (batch_count_.SetTotal(total_batches_.load())) {
+        finished_.MarkFinished();
+      }
+    }
```

Comment:
I feel like this `total_batches_` is redundant. Cant we do this instead?
```suggestion
    batch_count_.Increment(num_total);

    if (input_count_.Increment()) {
      outputs_[0]->InputFinished(this, total_batches_.load());
      int total_batches = batch_count_.count();
      if (batch_count_.SetTotal(total_batches)) {
        finished_.MarkFinished();
      }
    }
```

Reply:
you mean:
```cpp
if (input_count_.Increment()) {
      int total_batches = batch_count_.count();
      outputs_[0]->InputFinished(this, total_batches);
      if (batch_count_.SetTotal(total_batches)) {
        finished_.MarkFinished();
      }
    }
```
right?

Reply:
Please ignore this suggestion. This leads to a logic error as pointed out by @lidavidm